File: algorithms/1249. Minimum Remove to Make Valid Parentheses/solution.cc
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        // we can modify s because it is not a reference
        // learn from @votrubac
        std::stack<int> pindex;
        for (int i = 0; i < s.size(); ++i) {
            if ('(' == s[i]) {
                pindex.push(i);
            } else if (')' == s[i]) {
                if (pindex.empty()) {
                    s[i] = '.';
                } else {
                    pindex.pop();
                }
            }
        }
        while (pindex.size()) {
            s[pindex.top()] = '.';
            pindex.pop();
        }
        s.erase(remove(s.begin(), s.end(), '.'), s.end());
        return s;
    }
};

// Runtime: 24 ms, faster than 95.00% of C++ online submissions for Minimum Remove to Make Valid Parentheses.
// Memory Usage: 10.6 MB, less than 100.00% of C++ online submissions for Minimum Remove to Make Valid Parentheses.

File: algorithms/1249. Minimum Remove to Make Valid Parentheses/solution.py
class Solution(object):
    def minRemoveToMakeValid(self, s):
        """
        :type s: str
        :rtype: str
        """
        invalid_index = set()
        stack = []
        for i, c in enumerate(s):
            if c == "(":
                stack.append(i)
            elif c == ")":
                if not stack:
                    invalid_index.add(i)
                else:
                    stack.pop()
        invalid_index |= set(stack)
        res = []
        for i, c in enumerate(s):
            if i not in invalid_index:
                res.append(c)
        return ''.join(res)

# Runtime: 164 ms, faster than 84.76% of Python online submissions for Minimum Remove to Make Valid Parentheses.
# Memory Usage: 15.1 MB, less than 100.00% of Python online submissions for Minimum Remove to Make Valid Parentheses.
